### ISSUE TITLE
After async introduction -> Error communicating with OpenAI

### ISSUE THREAD

**Issue by gsext** [whitead/paper-qa#84]
After async feature introduction I am unable to run any query. I get below error with increasing waiting time and then Error communicating with OpenAI. 

By calling the chains with .run() instead of .arun() and out of async functions, I don't get the error.

Any suggestions on how to fix this issue? Thanks

`Retrying langchain.chat_models.openai.acompletion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI.`

**Comment by recursionbane** [comment]
I'm seeing something similar, but for locally-hosted LLMs that do not have Async generation implemented. It'd be nice to have synchronous generation as an option without having to downgrade to v0.1.0 (which does not allow for a local embeddings model).

**Comment by whitead** [comment]
#92 adds fallback so that LLMs that do not have async will switch to sync in chains

**Comment by recursionbane** [comment]
Not sure if this is resolved. Do we need an additional setting to enable async->sync fallback?

```
> pip list | egrep 'paper|langchain'
langchain                     0.0.157
paper-qa                      1.4.0
```

```
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                                                                                                         
  File "/opt/conda/lib/python3.8/site-packages/paperqa/qaprompts.py", line 91, in agenerate
    return await super().agenerate(input_list, run_manager=run_manager)
  File "/opt/conda/lib/python3.8/site-packages/langchain/chains/llm.py", line 90, in agenerate
    return await self.llm.agenerate_prompt(
  File "/opt/conda/lib/python3.8/site-packages/langchain/llms/base.py", line 136, in agenerate_prompt
    return await self.agenerate(prompt_strings, stop=stop, callbacks=callbacks)
  File "/opt/conda/lib/python3.8/site-packages/langchain/llms/base.py", line 266, in agenerate
    raise e
  File "/opt/conda/lib/python3.8/site-packages/langchain/llms/base.py", line 258, in agenerate
    await self._agenerate(
  File "/opt/conda/lib/python3.8/site-packages/langchain/llms/base.py", line 395, in _agenerate
    await self._acall(prompt, stop=stop, run_manager=run_manager)
  File "/opt/conda/lib/python3.8/site-packages/langchain/llms/base.py", line 363, in _acall
    raise NotImplementedError("Async generation not implemented for this LLM.")
NotImplementedError: Async generation not implemented for this LLM.
```

**Comment by whitead** [comment]
I'm very confused, the line [your stack trace is showing](https://github.com/whitead/paper-qa/blob/89c8ea8137127f18a0e12189350a72b9abe503ff/paperqa/qaprompts.py#L91) is wrapped in a `try/except` for a `NotImplementedError`. I have no idea how that could happen.

**Comment by hshany** [comment]
> 

I encountered the same problem and in my case the culprit was with the proxy I used to access openai service (yes I have to use one). The openai APIs use `requests` to handle sync requests and `aiohttp` to handle async requests. Whereas `requests` works well with socks5 proxies, `aiohttp` doesn't. I switched from a socks proxy to an http proxy and everything is working again. This may not be your case but hope it helps someone who ended up here.
